File: javascript/src/sliding-window/subarray-product-less-than-k.js
import { parseAssignments } from "../shared/io.js";
export const implemented = true;
export function solve(input) {
  const { nums, k } = parseAssignments(input);
  if (k <= 1) return 0;
  let left = 0, product = 1, count = 0;
  for (let right = 0; right < nums.length; right++) {
    product *= nums[right];
    while (product >= k) {
      product /= nums[left];
      left++;
    }
    count += right - left + 1;
  }
  return count;
}

File: java/src/main/java/algolens/slidingwindow/SubarrayProductLessThanK.java
package algolens.slidingwindow;

public class SubarrayProductLessThanK {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        if (k <= 1) return 0;
        int l = 0, count = 0;
        int product = 1;
        for (int r = 0; r < nums.length; r++) {
            product *= nums[r];
            while (product >= k) {
                product /= nums[l];
                l++;
            }
            count += r - l + 1;
        }
        return count;
    }
}
